fem/mgbilinearform.hpp

```diff
@@ -0,0 +1,43 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+#ifndef MFEM_MGBILINEARFORM
+#define MFEM_MGBILINEARFORM
+
+#include "../linalg/multigrid.hpp"
+#include "bilinearform.hpp"
+#include "spacehierarchy.hpp"
+
+namespace mfem
+{
+
+class MultigridBilinearForm : public MultigridOperator
```

Comment:
The reason I decided to implement a `MultigridBilinearForm` was that the user can easily create a new `MultigridOperator` with sane default smoothers and a reasonable coarse grid solver. This class also handles the memory of the forms which are required for each level in the hierarchy. In this way, the user only needs to define a bilinear form on the finest level and to provide a `SpaceHierarchy`. The bilinear forms for the other levels are being created automatically for all spaces in the `SpaceHierarchy` (that is what I need the newly introduced copy method in the integrators for).

I discussed with Veselin once, that we want the `MultigridOperator` class to be a linear algebra object which should not "know" anything about BilinearForms. This would be violated when adding the "named constructor" to `MultigridOperator`.

Do you think it would be okay to move what's inside of `MultigridBilinearForm` to the example file? Then the user has to do more work but can choose between different smoothers and coarse grid solvers.

Reply:
I might be the one of the few who actually likes the idea of the `MultigridBilinearForm`. I agree with Veselin's perspective that the operators should be agnostic about specifics like bilinear forms.

If there is projected generic future use of this class, it's better to remain in the library. If not, the compromise to move it to the example seems reasonable. If in the future it becomes necessary it can be pulled out of the example back to the library.

Reply:
I decided to move the (Par)MultigridBilinearForm classes into the respective examples. In the current state, these classes were too inflexible for the general library, as they are too application specific. I do think, however, that adding such class in the future may be reasonable.